pkg/provisioner/ironic/ironic.go

```diff
@@ -334,54 +416,51 @@ func (p *ironicProvisioner) InspectHardware() (result provisioner.Result, err er
-	// hardware details struct as part of a second pass.
-	if p.host.Status.HardwareDetails == nil {
-		p.log.Info("continuing inspection by setting details")
-		p.host.Status.HardwareDetails =
-			&metalkubev1alpha1.HardwareDetails{
-				RAMGiB: 128,
-				NIC: []metalkubev1alpha1.NIC{
-					metalkubev1alpha1.NIC{
-						Name:      "nic-1",
-						Model:     "virt-io",
-						Network:   "Pod Networking",
-						MAC:       "some:mac:address",
-						IP:        "192.168.100.1",
-						SpeedGbps: 1,
-					},
-					metalkubev1alpha1.NIC{
-						Name:      "nic-2",
-						Model:     "e1000",
-						Network:   "Pod Networking",
-						MAC:       "some:other:mac:address",
-						IP:        "192.168.100.2",
-						SpeedGbps: 1,
-					},
-				},
-				Storage: []metalkubev1alpha1.Storage{
-					metalkubev1alpha1.Storage{
-						Name:    "disk-1 (boot)",
-						Type:    "SSD",
-						SizeGiB: 1024 * 93,
-						Model:   "Dell CFJ61",
-					},
-					metalkubev1alpha1.Storage{
-						Name:    "disk-2",
-						Type:    "SSD",
-						SizeGiB: 1024 * 93,
-						Model:   "Dell CFJ61",
-					},
-				},
-				CPUs: []metalkubev1alpha1.CPU{
-					metalkubev1alpha1.CPU{
-						Type:     "x86",
-						SpeedGHz: 3,
-					},
-				},
+	status, err := introspection.GetIntrospectionStatus(p.inspector, ironicNode.UUID).Extract()
+	if err != nil {
+		if _, isNotFound := err.(gophercloud.ErrDefault404); isNotFound {
+			p.log.Info("starting new hardware inspection")
+			manageBoot := true
+			start := introspection.StartIntrospection(p.inspector,
+				ironicNode.UUID,
+				introspection.StartOpts{ManageBoot: &manageBoot})
+			if err := start.ExtractErr(); err != nil {
+				return result, errors.Wrap(err, "failed to begin hardware inspection")
 			}
+			result.Dirty = true
+			return result, nil
+		}
+		return result, errors.Wrap(err, "failed to extract hardware inspection status")
+	}
+	if !status.Finished {
+		p.log.Info("inspection in progress", "started_at", status.StartedAt)
+		result.Dirty = true // make sure we check back
+		result.RequeueAfter = introspectionRequeueDelay
+		return result, nil
+	}
+	if status.Error != "" {
+		p.log.Info("inspection failed", "error", status.Error)
+		p.publisher("InspectionFailed",
+			fmt.Sprintf("Host introspection failed: %s", status.Error))
+		result.Dirty = p.host.SetErrorMessage(status.Error)
+		return result, nil
+	}
+
+	// Introspection is ongoing
+	if p.host.Status.HardwareDetails == nil {
+		p.log.Info("getting hardware details from inspection")
+		introData := introspection.GetIntrospectionData(p.inspector, ironicNode.UUID)
+		data, err := introData.Extract()
+		if err != nil {
+			return result, errors.Wrap(err, "failed to retrieve hardware introspection data")
+		}
+		if jsonData, err := json.Marshal(introData.Body); err == nil {
+			p.log.Info("Received introspection data", "data", jsonData)
+		} else {
+			p.log.Info("Error marshalling JSON introspection data", "error", err)
```

Comment:
Should this error case return right away?

Reply:
It's only an error in the logging, so I'd say no.

Reply:
It would be an error from json.Marshal(), right?

Reply:
Yes, but we're marshalling it purely so that we can log the data in JSON format. Which we need to do purely because gophercloud stores the body as deserialised data, even though it was received as serialised JSON. IOW it can't fail and even if it did, it only means that you wouldn't get to see the log message.

I think this whole JSON serialisation thing might actually be a mistake though, because the logging code seems to base64 encode the JSON. It is a structured logging thing, so maybe it would let us just pass the deserialised objects?

Reply:
Ah,  I see, I was not reading the variable names closely and saw the call to getHardwareDetails() a few lines down and assumed it was using the results of the Marshall call.

Yes, I think if you pass the structured object to the log call it will figure out how to do the right thing with it. That may be quite a lot of data, though. Do we really want to log all of it in one go?

Reply:
It is a ton of data :/ I was tempted to take the log out before posting the PR, but this will likely be invaluable for debugging - every time somebody installs on some new weird hardware and it doesn't show up the way we want, we'll be able to just ask them for the log.

Reply:
What if we log the sections as we try to interpret them? "Here's all the CPU info we found: blob", etc.?

Reply:
I'll look into that. The data is quite spread around though, so not sure if it is feasible.

Reply:
Is there a way to have ironic log it in its own logs? That would at least keep the "noise" separate from this controller.

Or we could write it out to disk in the container and just ask someone to exec into the pod if they really need the data.

Reply:
Let's leave it in the operator logs for now. I know we have more data to pull out of the structure, and it's going to make development easier. We can clean it up before we go GA.

Reply:
+1 for clean it up later